### PR TITLE
[wip] util: Improve evaluation of includeconf lines in network sections of the config file

### PR DESCRIPTION
**Pull request by AkioNak**
This PR intends to make it easy to understand how the configuration files are interpreted.

1. Evaluate "includeconf" at the position described in the config file (like #include directive in C/C++), rather than at the end.

2. If once a section is specified with square brackets like [main], the only way to configure another network is to use square brackets again, e.g. [test].

3. In the "included" config file, the section is taken over from the "including" file. If no sections are specified yet, using square brackets or section prefix is still possible in the "included" file.

4. Changing the section in the "included" config file does not affects the including file. The section of the including file is kept.

5. If the "included" config file is included after the section has been specified by using square brackets or if they are included by using section prefix, any square brackets can not be used in that file.

ex) With Following 3 files, the properties are read as: 
```
    port=8444, main.port=8445, test.port=8442, regtest.port: undefined
```

 bitcoin.conf
 ---
 ```
test.includeconf=inc1.conf
includeconf=inc2.conf
port=8441   # ignored (2nd appearance of w/o section)
```

 inc1.conf
 ---
```
 port=8442   # test
 [regtest]
 port=8443   # ignored
```

 inc2.conf
 ---
```
 port=8444   # w/o section
 [main]
 port=8445   # main
 [regtest]
```

### CODE REVIEW

**Comment by promag** [comment]
This can break existing configurations correct?

**Comment by AkioNak** [comment]
@promag Thanks for your comment.
Yes. Sometimes this may be different from previous results when interpreting existing configurations.
Especially the possibility of interpretation unmatch increases if using "includeconf" and sections are specified with square brackets in the "included" configuration file.
Is it necessary for this change to add a flag to specify how to read the configuration file?

File: src/util/system.cpp [hunk 1]
```diff
@@ -856,22 +857,50 @@ static bool GetConfigOptions(std::istream& stream, std::string& error, std::vect
...
                     return false;
                 }
-                options.emplace_back(name, value);
-                if ((pos = name.rfind('.')) != std::string::npos) {
-                    sections.insert(name.substr(0, pos));
+                options.emplace_back(name.empty() ? key : name + '.' + key, value);
+                if (key == "includeconf") {
+                    if (depth < 0) {
+                        continue;
+                    } else if (depth < MAX_INCLUDECONF_DEPTH) {
```
Comment (practicalswift) [inline src/util/system.cpp]: Nit: The `else` here is redundant.

**Comment by promag** [comment]
You should not break existing configurations, especially if there is no strong reason to do it.

File: src/util/system.cpp (hunk 1 above)
Comment (AkioNak) [reply src/util/system.cpp]: Done.

**Comment by dougstrong77** [comment]
I need advise on some things that accurd

On Dec 4, 2018 6:40 AM, "João Barbosa" <notifications@github.com> wrote:

This can break existing configurations correct?

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub
<https://github.com/bitcoin/bitcoin/pull/14866#issuecomment-444122125>, or mute
the thread

**Comment by laanwj** [comment]
> @promag The includeconf and sections features are both relatively recent

I think that's a good point, does this *only* affect use of `-includeconf`?

**Comment by kallewoof** [comment]
@laanwj I believe so, yes, except for maybe the minor detail (item 4 in the OP list). @AkioNak?

**Comment by AkioNak** [comment]
Rebased but travis failed. Hmm,,, I will check what's going on.

**Comment by AkioNak** [comment]
The direct cause of failing travis was potential deadlock detection. 
The investigation will continue ...

potential_deadlock_detected() reports following message:
```
POTENTIAL DEADLOCK DETECTED
Previous lock order was:
 (1) cs_args  util/system.cpp:921
 (2) csPathCached  util/system.cpp:790
Current lock order is:
 (2) csPathCached  util/system.cpp:790
 (1) cs_args  util/system.cpp:531
```

**Comment by AkioNak** [comment]
Addressed to the potential deadlock by narrowing the range of LOCK(cs_args).

**Comment by kallewoof** [comment]
utACK cca8a9196f491aa8e99d18846578307cd7004c45

**Comment by AkioNak** [comment]
squashed and rebased.

**Comment by Sjors** [comment]
Concept ACK, except I think the following example (from the description) should just fail:

```ini
[main]
includeconf=inc1.conf
test.includeconf=inc2.conf
```

Once a network is specified with square brackets like `[main]`, the only way to configure another network should be to use square brackets again, e.g. `[testsnet]`. That rule should also apply to anything included by `includeconf`.

It's important to simplify this behavior sufficiently so we stand a chance of one day merging #11082 :-)

**Comment by AkioNak** [comment]
@Sjors Thank you for your surggestions.

> Once a network is specified with square brackets like `[main]`, the only way to configure another network should be to use square brackets again, e.g. `[testsnet]`. That rule should also apply to anything included by `includeconf`.

Good idea. I agree.

> Or, slightly less strict, that square brackets may not exist in `includeconf` files if they are included after a square bracket.

I like this. And I think square brackets may not exist in the files that are included by using "section.includeconf" style, too.

**Comment by Sjors** [comment]
Ok, looking forward to the new version. I think it would be very useful to add a lot more tests to [feature_config_args.py](https://github.com/bitcoin/bitcoin/blob/master/test/functional/feature_config_args.py), given that you were able to make these changes without breaking any existing test.

**Comment by AkioNak** [comment]
rebased. (no functional change yet)

**Comment by AkioNak** [comment]
> Once a network is specified with square brackets like [main], the only way to configure another network should be to use square brackets again, e.g. [testsnet]. That rule should also apply to anything included by includeconf.

Added rules for how to use square brackets & preriods.
Also added some tests.

@Sjors Could you review this?

**Comment by AkioNak** [comment]
Squashed.
(Also updated commit log)

**Comment by AkioNak** [comment]
Rebased.

**Comment by AkioNak** [comment]
Rebased.

**Comment by jtimon** [comment]
I don't like that in the example port=8441 in bitcoin.conf is applied only to regtest just because inc2.conf ended with [regtest].
I feel like I could get crazy for a few hours just by forgetting inc2.conf ends with [regtest] if, let's say, it previously ended with [main] but I moved chain sections around in inc.conf or something.

Looking at the example, I just can't avoid wondering why the user can't just add [regtest] to bitcoin.conf like this:

```
test.includeconf=inc1.conf
includeconf=inc2.conf
[regtest]
port=8441   # regtest
```

or just put the regtest port in inc2.conf like this.

```
port=8444   # w/o section
[main]
port=8445   # main
[regtest]
port=8441   # regtest
```

I've personally never used includeconf myself, so perhaps I just don't understand how people are using this feature.
 
Perhaps this can be done in another PR, but I feel like after includeconf, you should go back to whatever section you were in bitcoin.conf, not remain in the last one set by the included file.

**Comment by AkioNak** [comment]
@jtimon Thank you  for the comment!

> after includeconf, you should go back to whatever section you were in bitcoin.conf, not remain in the last one set by the included file.

It makes sense. That behavior is natural so I will address.

**Comment by AkioNak** [comment]
@jtimon I have addressed.
(Changing section in the included file never affect to the including file.)

**Comment by AkioNak** [comment]
Rebased.

**Comment by Sjors** [comment]
Concept re-ACK. The latest set of 5 criteria in the PR descriptor look very reasonable.

Needs rebase after @ryanofsky's settings merge refactor in #15934 (which also added lots of useful tests). I think the increased sanity warrants (if needed) breaking backwards compatibility.

**Comment by ryanofsky** [comment]
Concept ACK, I think. But I am having a trouble understanding the description, because it seems to be describing a lot of current behavior (like points 2 and 4) that the PR doesn't actually affect, and not be clear about what is changing.

Would the following be an accurate description of what the PR does?

> Included configurations from `includeconf` lines in `[main]` `[test]` and `[regtest]` sections of the config file are now evaluated in context of that section and not treated like top-level includes. Config files included within a section are also now disallowed from changing settings in other sections.

If that is accurate, it would also be helpful to add it as an entry to: https://github.com/bitcoin/bitcoin/blob/master/doc/release-notes.md

The documentation https://github.com/bitcoin/bitcoin/blob/master/doc/bitcoin-conf.md should also be updated. I think most of the text in the current PR description should be removed and incorporated into that file instead.

I would also change the title of the PR to better describe what it is doing. Instead of "Improve property evaluation way in bitcoin.conf" maybe "util: Improve evaluation of includeconf lines in network sections of the config file"

**Comment by AkioNak** [comment]
@Sjors @ryanofsky Thank you for your comments/suggestions.
I will rebase and also address the suggestions above.

**Comment by adamjonas** [comment]
@AkioNak pinging for a rebase and addressing the comments by ryanofsky and Sjors.

**Comment by AkioNak** [comment]
@adamjonas I will do.

**Comment by AkioNak** [comment]
Rebased.
I am still working on addressing the comments by ryanofsky and Sjors.

**Comment by adamjonas** [comment]
Hi @AkioNak - friendly monthly ping.

**Comment by AkioNak** [comment]
@adamjonas Thanks. I will do again.

**Comment by AkioNak** [comment]
I will solve the conflicts.

**Comment by fanquake** [comment]
Going to close with "Up for Grabs".

**Pull request closed without merging**